endpoints/s3/src/putobject.cpp

```diff
@@ -53,40 +53,63 @@ void manually_parse_chunked_body_write_to_irods_in_background(
 {
     irods::http::session_pointer_type session_ptr_;
     beast::http::response<beast::http::empty_body> resp_;
     std::shared_ptr<beast::http::request_parser<boost::beast::http::buffer_body>> parser_;
+    std::string irods_path_;
+    bool upload_part_flag_;
     std::string part_filename_;
     std::ofstream part_file_;
+    irods::experimental::io::odstream odstream_;
     std::vector<char> buffer_;
     std::size_t total_bytes_read_{};
-
+    std::shared_ptr<irods::experimental::client_connection> conn_;
+    irods::experimental::io::client::default_transport tp_;
   public:
     incremental_async_read(
         std::shared_ptr<beast::http::request_parser<boost::beast::http::buffer_body>>& _parser,
         irods::http::session_pointer_type& _session_ptr,
         beast::http::response<beast::http::empty_body>& _response,
-        std::string _etag,
-        std::string _part_filename)
+        std::string _irods_path,
+        bool _upload_part_flag,
+        std::string _part_filename,
+        std::shared_ptr<irods::experimental::client_connection> _conn)
         : session_ptr_{_session_ptr->shared_from_this()}
         , resp_{std::move(_response)}
         , parser_{_parser}
+        , irods_path_{_irods_path}
+        , upload_part_flag_{_upload_part_flag}
         , part_filename_{std::move(_part_filename)}
         , buffer_(irods::s3::get_put_object_buffer_size_in_bytes())
+        , conn_{_conn}
+        , tp_{*conn_}
     {
         resp_.version(parser_->get().version());
-        resp_.set("Etag", _etag);
+        resp_.set("Etag", _irods_path);
         resp_.keep_alive(parser_->get().keep_alive());
 
-        logging::trace("{}: Opening part file [{}] for writing.", __func__, part_filename_);
-        part_file_.open(part_filename_);
-        if (!part_file_) {
-            auto msg = fmt::format("{}: Failed to open part file for writing [{}].", __func__, _part_filename);
-            logging::error(msg);
-            THROW(SYS_INTERNAL_ERR, std::move(msg));
+        if (upload_part_flag_) {
+            logging::trace("{}: Opening part file [{}] for writing.", __func__, part_filename_);
+            part_file_.open(part_filename_);
+            if (!part_file_) {
+                auto msg = fmt::format("{}: Failed to open part file for writing [{}].", __func__, _part_filename);
+                logging::error(msg);
+                THROW(SYS_INTERNAL_ERR, std::move(msg));
+            }
+        } else {
+            logging::trace("{}: Opening iRODS file [{}] for writing.", __func__, part_filename_);
+            odstream_.open(tp_, std::move(irods_path_), irods::experimental::io::root_resource_name{irods::s3::get_resource()});
+            if (!odstream_) {
+                auto msg = fmt::format("{}: Failed to open irods file for writing [{}].", __func__, _irods_path);
+                logging::error(msg);
+                THROW(SYS_INTERNAL_ERR, std::move(msg));
+            }
```

Comment:
I'm confused as to why this is needed? Is it for something other than multipart upload?

Uploads with and without multipart enabled worked before this change.

Reply:
This is for both multipart and non-multipart.  The uploads without multipart (with boost http parsing the body) were temporarily broken .  I had to update the code to make it work.

Remember there are four possibilities here:

1. Regular upload with boost http parsing the body.
2. Multipart upload with boost http parsing the body.
3. Regular upload with manually parsing of chunk body.
4. Multipart upload with manually parsing of chunk body.

Reply:
`mc` worked in all cases. `aws` worked in the multipart case _(that's all I tested)_.

I'm guessing (1) for the `aws` client is the culprit. Is that correct?